src/software/sensor_fusion/sensor_fusion_test.cpp

```diff
@@ -875,3 +875,115 @@ TEST_F(SensorFusionTest, test_sensor_fusion_reset_behaviour_ignore_bad_packets)
     result = *sensor_fusion.getWorld();
     EXPECT_EQ(initWorld(), result);
 }
+
+TEST_F(SensorFusionTest, breakbeam_pass_update_test)
+{
+    // The following code test this thing: if the breakbeam is triggered and the position
+    // of the robot and the ball given by the ssl vision system is less than a certain
+    // threshold, we expect shouldUseRobotBallPositionInsteadOfSSL to return True
+    SensorProto sensor_msg;
+
+    // creating ssl vision
+    const uint32_t camera_id    = 0;
+    const uint32_t frame_number = 40391;
+    Timestamp time              = Timestamp::fromSeconds(12.1);
+
+    Point ssl_position(0.05, 0.05);
+    Vector velocity(0, 0);
+    ball_state = BallState{ssl_position, velocity};
+
+    yellow_robot_states.clear();
+    blue_robot_states.clear();
+
+    RobotState robot_state(Point(0, 0), Vector(0, 0), Angle::fromRadians(2),
+                           AngularVelocity::zero());
+    RobotStateWithId robot_id = {2, robot_state};
+    yellow_robot_states.push_back(robot_id);
+
+    std::unique_ptr<SSLProto::SSL_DetectionFrame> frame =
+        createSSLDetectionFrame(camera_id, time, frame_number, {ball_state},
+                                yellow_robot_states, blue_robot_states);
+
+    // creating robot status
+    auto robot_msg = std::make_unique<TbotsProto::RobotStatus>();
+    robot_msg->set_robot_id(2);
+
+    auto power_status_msg = std::make_unique<TbotsProto::PowerStatus>();
+    power_status_msg->set_breakbeam_tripped(true);
+    *(robot_msg->mutable_power_status()) = *power_status_msg;
+
+    // create ssl wrapper packet
+    auto gemotry_data = initSSLDivBGeomData();
+    auto ssl_wrapper_packet =
+        createSSLWrapperPacket(std::move(gemotry_data), std::move(frame));
+    *(sensor_msg.mutable_ssl_vision_msg()) = *ssl_wrapper_packet;
+    *(sensor_msg.add_robot_status_msgs())  = *robot_msg;
+
+    // we have to call update sensor fusion two times because of updating ball position
+    // comes before updating breakbeam tripped. Thus, we have to call this function two
+    // times before the ball position is updated
+    sensor_fusion.processSensorProto(sensor_msg);
+    sensor_fusion.processSensorProto(sensor_msg);
+
+    Point updated_ball_pos = sensor_fusion.getWorld().value().ball().position();
+
+    // is it using the robot state position
+    EXPECT_TRUE(updated_ball_pos != robot_state.position());
+    // is it not using the ssl vision?
+    EXPECT_TRUE(updated_ball_pos == ssl_position);
```

Comment:
these assertions look contradictory to what the test is trying to do

Reply:
indeed.